crates/stark-felt/src/lib.rs

```diff
@@ -398,6 +691,150 @@ mod arithmetic {
         }
     }
 
+    impl ops::Shl<usize> for Felt {
+        type Output = Felt;
+        fn shl(self, rhs: usize) -> Self::Output {
+            Self(FieldElement::from(&self.0.representative().shl(rhs)))
+        }
+    }
+
+    impl ops::Shl<usize> for &Felt {
+        type Output = Felt;
+
+        fn shl(self, rhs: usize) -> Self::Output {
+            Felt(FieldElement::from(&self.0.representative().shl(rhs)))
+        }
+    }
+
+    impl ops::Shr<usize> for Felt {
+        type Output = Felt;
+        fn shr(self, rhs: usize) -> Self::Output {
+            Self(FieldElement::from(&self.0.representative().shr(rhs)))
+        }
+    }
+
+    impl ops::Shr<usize> for &Felt {
+        type Output = Felt;
+
+        fn shr(self, rhs: usize) -> Self::Output {
+            Felt(FieldElement::from(&self.0.representative().shr(rhs)))
+        }
+    }
```

Comment:
It sounds like Montgomery repr improves mathematical operation to the detriment of bitwise one. But it's a tradeoff we are willing to do I think.

Reply:
I asked internally for the reason the cairo lib does not imp any bitshift operations. Maybe there is a good one and they will tell us :)

Reply:
> It sounds like Montgomery repr improves mathematical operation to the detriment of bitwise one. But it's a tradeoff we are willing to do I think.

Indeed. The mention of it is mostly that unless something definitely depends on bitwise operations on felts, it might be better to fix the VM to not use them than implement them here, specially if it was avoided on purpose in the lib.

Reply:
Maybe, because bitshift can be thought are multiplication and division by 2^n, it can be left out.

Reply:
Generally - as a rule - i won't implement any of the bit based operations here - unless i have a specific performance reason to implement it (but in any case, another method not for this impl would still work).|
IMO - felt types bit lookups don't make sense, as the type is not aligned to bits, and overflowing operations would fail in a non-trivial manner.
generally - i don't like to have operators for operations that don't clearly mean something specific.
For shifting and such i would rather explicitly back and forth to a BigInt as there i understand exactly what the operation mean -  and i know that the failure is caused by the conversion, and not the operation.

Reply:
Thanks @orizi!

@Oppen would you be ok with removing all the bitwise operations from this lib?

And more generally cement the idea that we should not be using `representative` while implementing Felt only operations

Reply:
> Maybe, because bitshift can be thought of as multiplication and division by 2^n, it can be left out.

Yes, and that's why it was used. Back when we used `num_bigint` for felts in `cairo-vm` replacing those multiplications and divisions by shifts was a very significant optimization. With Montgomery that becomes something between irrelevant and harmful.